beefy-gadget/src/gossip.rs

```diff
@@ -53,41 +49,47 @@ where
 /// rejected/expired.
 ///
 ///All messaging is handled in a single BEEFY global topic.
-pub(crate) struct BeefyGossipValidator<B>
+pub(crate) struct GossipValidator<B>
 where
 	B: Block,
 {
 	topic: B::Hash,
 	live_rounds: RwLock<Vec<NumberFor<B>>>,
 }
 
-impl<B> BeefyGossipValidator<B>
+impl<B> GossipValidator<B>
 where
 	B: Block,
 {
-	pub fn new() -> BeefyGossipValidator<B> {
-		BeefyGossipValidator {
+	pub fn new() -> GossipValidator<B> {
+		GossipValidator {
 			topic: topic::<B>(),
 			live_rounds: RwLock::new(Vec::new()),
 		}
 	}
 
-	/// Note a live voting round
+	/// Note a voting round.
+	///
+	/// Noting `round` will keep `round` live.
 	///
-	/// This should be called, in order to keep tabs on `round`.
+	/// We retain the [`MAX_LIVE_GOSSIP_ROUNDS`] most **recent** voting rounds as live.
+	/// As long as a voting round is live, it will be gossiped to peer nodes.
 	pub(crate) fn note_round(&self, round: NumberFor<B>) {
 		trace!(target: "beefy", "🥩 About to note round #{}", round);
 
-		let mut live_rounds = self.live_rounds.write();
+		let mut live = self.live_rounds.write();
 
-		// NOTE: ideally we'd use a VecDeque here, but currently binary search is only available on
-		// nightly for `VecDeque`.
-		while live_rounds.len() > MAX_LIVE_GOSSIP_ROUNDS {
-			let _ = live_rounds.remove(0);
+		// we want to keep `MAX_LIVE_GOSSIP_ROUNDS` most recent rounds.
+		if (live.len() == MAX_LIVE_GOSSIP_ROUNDS) && (live[0] > round) {
+			return;
 		}
```

Comment:
I feel this condition is superfluous now.

Reply:
I could now start to argue that the early return is a bit of an optimization. But that is BS, given the number of live rounds we are going to keep around. So, guilty as charged. I just missed it :angry:

Reply:
Yeah, I was thinking the same initially, but then suggested this simplified code and I think less code will be much more important than an extra binary search on 5-element collection :)